File: Factory Method/Exercicios/Ex1_Log/Ex1.py
"""
PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)

Exercício: Sistema de Logging.
Este padrão é ideal para sistemas de Log onde a forma de registro 
pode mudar dependendo do ambiente (Desenvolvimento vs Produção).
"""

from abc import ABC, abstractmethod

# 1. PRODUTO ABSTRATO (Interface)
class Logger(ABC):
    @abstractmethod
    def log(self, mensagem):
        pass

# 2. PRODUTOS CONCRETOS
class ConsoleLogger(Logger):
    def log(self, mensagem):
        print(f"[CONSOLE LOG]: {mensagem}")

class FileLogger(Logger):
    def log(self, mensagem):
        print(f"[ARQUIVO LOG]: Registrando no arquivo oculto -> {mensagem}")

# 3. CRIADOR (Classe Base)
class LogManager(ABC):
    @abstractmethod
    def create_logger(self) -> Logger:
        """Este é o Factory Method."""
        pass

    def registrar(self, mensagem):
        """Método que utiliza o Logger criado pela subclasse."""
        logger = self.create_logger()
        logger.log(mensagem)

# 4. CRIADORES CONCRETOS (Fábricas)
class ConsoleLogManager(LogManager):
    def create_logger(self):
        return ConsoleLogger()

class FileLogManager(LogManager):
    def create_logger(self):
        return FileLogger()

# 5. CLIENTE
if __name__ == "__main__":
    c_log = ConsoleLogManager()
    c_log.registrar("Otimização de consulta iniciada.")

    f_log = FileLogManager()
    f_log.registrar("Aviso: Baixa memória disponível no servidor.")

File: Factory Method/Exercicios/Ex1_Log/Ex1.java
/**
 * PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)
 * 
 * Exercício: Sistema de Logging.
 * O objetivo é permitir que o sistema registre mensagens (Logs) 
 * sem saber para onde essas mensagens estão sendo enviadas (Console, Arquivo, etc).
 */

// 1. PRODUTO ABSTRATO
interface Logger {
    void log(String mensagem);
}

// 2. PRODUTOS CONCRETOS
class ConsoleLogger implements Logger {
    public void log(String mensagem) {
        System.out.println("[CONSOLE LOG]: " + mensagem);
    }
}

class FileLogger implements Logger {
    public void log(String mensagem) {
        System.out.println("[ARQUIVO LOG]: Gravando no arquivo log.txt -> " + mensagem);
    }
}

// 3. CRIADOR ABSTRATO (Define o Factory Method)
abstract class LogManager {
    // Este é o Factory Method
    public abstract Logger createLogger();

    // Método que utiliza o produto criado pelo Factory Method
    public void registrar(String mensagem) {
        Logger logger = createLogger();
        logger.log(mensagem);
    }
}

// 4. CRIADORES CONCRETOS
class ConsoleLogManager extends LogManager {
    @Override
    public Logger createLogger() {
        return new ConsoleLogger();
    }
}

class FileLogManager extends LogManager {
    @Override
    public Logger createLogger() {
        return new FileLogger();
    }
}

// 5. CLIENTE
public class Ex1 {
    public static void main(String[] args) {
        LogManager console = new ConsoleLogManager();
        console.registrar("Sistema iniciado com sucesso.");

        LogManager arquivo = new FileLogManager();
        arquivo.registrar("Erro detectado na linha 42.");
    }
}
